kythe/docs/schema/indexing-protobuf.txt

```diff
@@ -31,12 +31,235 @@ implementation, as this is often difficult (or uninteresting) to read. This
+  RED;
+}
+-------------------------------------------------------------------------------
+
+Generator produces `color.js`:
+[source,javascript]
+-------------------------------------------------------------------------------
+const Color = {
+  RED: 0,
+};
+-------------------------------------------------------------------------------
+
+==== Changes to Generator
+
+To support cross-references betwen `color.js` and `Color.java` we need to update
+Generator to output the following mapping data for `Color`, `RED` and `GREEN`
+elements.
+
+[source,json]
+-------------------------------------------------------------------------------
+{
+    "type": "kythe0",
+    "meta": [{
+        "type": "anchor_anchor",
+        "source_begin": 13,
+        "source_end": 18,
+        "target_begin": 6,
+        "target_end": 11,
+        "edge": "/kythe/edge/imputes",
+        "source_vname": {
+            "corpus": "corpus",
+            "path": "path/to/Color.java"
+        }
+    }, {
+        "type": "anchor_anchor",
+        "source_begin": 22,
+        "source_end": 25,
+        "target_begin": 18,
+        "target_end": 21,
+        "edge": "/kythe/edge/imputes",
+        "source_vname": {
+            "corpus": "corpus",
+            "path": "path/to/Color.java"
+        }
+    }]
+}
+-------------------------------------------------------------------------------
+
+This mapping has 3 `meta` entries. The first entry for `Color`, the second for
+`RED`. Note:
+
+* Each entry doesn't contain names of elements. Each entry contains only
+  position of elements in the source (`Color.java`) and target (`color.js`)
+  files.
+* Each position is defined as byte offset inside file and not as line/column.
+  This is required because in Kythe anchors are defined using byte offsets and
+  not line/column. In this example JavaScript indexer will process this
+  mapping and will need to output *anchor* for `Color.java` and indexer
+  doesn't have access to the `Color.java` file (it has access only to JS
+  files). Because of that JS indexer can't translate line/column to byte
+  offset.
+* Entry doesn't contain vnames of elements in `Color.java` or `color.js` and
+  instead contains positions. Vnames of nodes are internal details of each
+  indexer and subject to change. Generator usually a standalone tool that
+  doesn't know rules for producing vnames for specific language so it's
+  impossible for Generator to output vnames of nodes.
+
+To pass this mapping to the JavaScript Indexer Generator will append it
+as a comment at the last line of `color.js`:
+
+[source,javascript]
+-------------------------------------------------------------------------------
+const Color = {
+  RED: 0,
+};
+
+// Kythe Indexing Metadata:
+// {"type":"kythe0","meta":[{"type":"anchor_anchor","source_begin":13,...
+-------------------------------------------------------------------------------
+
+Inlining metadata inside `color.js` has benefit of avoiding passing extra
```

Comment:
The indexer takes a range of text in the generated file.  When it emits a "defines/binding" edge covering that range, it also emits an edge from the semantic entity it just defined at that location towards the entity specified in the metadata file.  For "imputes" edges, the semantic entity points at an anchor. For "generates" edges the target language indexer emits a reverse edge to the semantic entity which created the text range.  Of note, the indexer in question *always* emits at least one semantic node whose VName is known only to it.

In both of these cases, the post-processor has to do some work:
For imputes edges, it has to emit the appropriate generates edge pair.
For generates edges, it has to emit the forward edge.

There is no way to feed data into post-processing which doesn't come from an indexer.

Reply:
So IIUC the inlining is for the indexer to emit the semantic VName of the `defines/binding` generated at the given span (annotated in the inline data). If VNames are not predictable, that sounds fair.

About having to emit the forward edge for `generates`: why can't the indexer (or an external process) emit the forward edge directly? It is just the source and target swapped, isn't it?

Reply:
Yes, I think it would be possible to pass anchor-to-anchor mapping from generator to post-processing and let post-processing to find two corresponding `defines/binding` anchors and stich semantic VNames together. But as Shahms mentioned there is no way to feed data outside of indexer. Personally it makes it a little bit cleaner as you can assume only indexers output data. 

As for forward edges for `generates` - I don't know. Maybe to keep indexer a little bit simpler?